image/image.go

```diff
@@ -0,0 +1,386 @@
+	model, err := decodeModelAssertion(opts.ModelAssertionFn)
+	if err != nil {
+		return err
+	}
+	if err := os.MkdirAll(opts.GadgetUnpackDir, 0755); err != nil {
+		return fmt.Errorf("cannot create gadget unpack dir %q: %s", opts.GadgetUnpackDir, err)
+	}
+
+	dlOpts := &downloadOptions{
+		TargetDir:    opts.GadgetUnpackDir,
+		Channel:      opts.Channel,
+		StoreID:      model.Store(),
+		Architecture: model.Architecture(),
+	}
+	snapFn, _, err := acquireSnap(model.Gadget(), dlOpts)
+	if err != nil {
+		return err
+	}
+	// FIXME: jumping through layers here, we need to make
+	//        unpack part of the container interface (again)
+	snap := squashfs.New(snapFn)
+	return snap.Unpack("*", opts.GadgetUnpackDir)
+}
+
+func acquireSnap(snapName string, dlOpts *downloadOptions) (string, *snap.Info, error) {
+	if osutil.FileExists(snapName) {
+		return copyLocalSnapFile(snapName, dlOpts.TargetDir)
+	}
+
+	return downloadSnapWithSideInfo(snapName, dlOpts)
+}
+
+func bootstrapToRootdir(opts *Options) error {
+	if opts.Rootdir != "" {
+		dirs.SetRootDir(opts.Rootdir)
+		defer dirs.SetRootDir("/")
+	}
+
+	// sanity check target
+	if osutil.FileExists(dirs.SnapStateFile) {
+		return fmt.Errorf("cannot bootstrap over existing system")
+	}
+
+	model, err := decodeModelAssertion(opts.ModelAssertionFn)
+	if err != nil {
+		return err
+	}
+
+	// put snaps in place
+	if err := os.MkdirAll(dirs.SnapBlobDir, 0755); err != nil {
+		return err
+	}
+
+	snapSeedDir := filepath.Join(dirs.SnapSeedDir, "snaps")
+	dlOpts := &downloadOptions{
+		TargetDir:    snapSeedDir,
+		Channel:      opts.Channel,
+		StoreID:      model.Store(),
+		Architecture: model.Architecture(),
+	}
+
+	// FIXME: support sideloading snaps by copying the boostrap.snaps
+	//        first and keeping track of the already downloaded names
+	snaps := []string{}
+	snaps = append(snaps, opts.Snaps...)
+	snaps = append(snaps, model.Gadget())
+	snaps = append(snaps, model.Core())
+	snaps = append(snaps, model.Kernel())
+	snaps = append(snaps, model.RequiredSnaps()...)
+
+	for _, d := range []string{dirs.SnapBlobDir, snapSeedDir} {
+		if err := os.MkdirAll(d, 0755); err != nil {
+			return err
+		}
+	}
+
+	var seedYaml snap.Seed
+	for _, snapName := range snaps {
+		fmt.Printf("Fetching %s\n", snapName)
+		fn, info, err := acquireSnap(snapName, dlOpts)
+		if err != nil {
+			return err
+		}
+		// kernel/os are required for booting
+		if snapName == model.Kernel() || snapName == model.Core() {
+			if err := osutil.CopyFile(fn, filepath.Join(dirs.SnapBlobDir, filepath.Base(fn)), 0); err != nil {
+				return err
+			}
+		}
+		// set seed.yaml
+		seedYaml.Snaps = append(seedYaml.Snaps, &snap.SeedSnap{
+			Name:        info.Name(),
+			SnapID:      info.SnapID,
+			Revision:    info.Revision,
+			Channel:     info.Channel,
+			DeveloperID: info.DeveloperID,
+			Developer:   info.Developer,
+			File:        filepath.Base(fn),
+		})
+	}
+	if err := seedYaml.Write(); err != nil {
```

Comment:
Feels slightly uneven to hit this without a location for the file to be written. We've been providing the path in for everything else above, I think?